src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -4398,7 +4399,6 @@ protected int notValidUtf8(Object string, int byteIndex,
     @Primitive(name = "string_character_index", lowerFixnum = 2)
     public abstract static class StringCharacterIndexPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
-        @TruffleBoundary
```

Comment:
The `.getBytes()` calls below need to use BytesNode:
https://github.com/oracle/truffleruby/pull/2380/checks?check_run_id=2854694035

Reply:
There are quire a few loops below.
These need LoopNode.reportLoopCount + LoopConditionProfile for conditions and loop exits (e.g., `if () break`).
This is necessary so:
* Truffle compiles this method early enough.
* Graal can compute the loop frequency accurately and decide e.g., whether to hoist some nodes out of the loop or to perform loop peeling, etc. There is an upcoming performance warning when LoopConditionProfile is not used, so we don't want to regress there.

Reply:
`stringRope.isSingleByteOptimizable()` needs to be replace by a node, otherwise the getCodeRange() inside is a virtual call, which leads to a perf warning and not being able to inline during PE.

Reply:
I've taken another look and can't see where `stringRope.isSingleByteOptimizable()` is being used. Part of the changes I made was to add `singleByteOptimizableNode` and use that in the guards. Originally, I was just using `stringRope.isSingleByteOptimizable()`, but I did see the perf warning and changed. I didn't realize `getCodeRange()` was virtual until I saw the warning. I identified a bunch of other places to go swap that out in, but I planned on doing that in a future PR.

I'll add the loop counters. I thought we really only used those for implementing guest-level loops.

Reply:
Oh, I see now. I somehow removed the boundary on `StringCharacterIndexPrimitiveNode`. I wasn't even working on that one. I was flipping between machines with force pushes though and probably messed that up along the way. I'll restore the boundary.